src/aosm/azext_aosm/_configuration.py

```diff
@@ -0,0 +1,771 @@
+
+    def path_from_cli_dir(self, path: str) -> str:
+        """
+        Convert path from config file to path from current directory.
+
+        We assume that the path supplied in the config file is relative to the
+        configuration file.  That isn't the same as the path relative to where ever the
+        CLI is being run from.  This function fixes that up.
+
+        :param path: The path relative to the config file.
+        """
+        assert self.config_file
+
+        # If no path has been supplied we shouldn't try to update it.
+        if path == "":
+            return ""
+
+        # If it is an absolute path then we don't need to monkey around with it.
+        if os.path.isabs(path):
+            return path
+
+        config_file_dir = Path(self.config_file).parent
+
+        updated_path = str(config_file_dir / path)
+
+        logger.debug("Updated path: %s", updated_path)
+
+        return updated_path
+
+    @property
+    def output_directory_for_build(self) -> Path:
+        """Base class method to ensure subclasses implement this function."""
+        raise NotImplementedError("Subclass must define property")
+
+    @property
+    def acr_manifest_names(self) -> List[str]:
+        """The list of ACR manifest names."""
+        raise NotImplementedError("Subclass must define property")
+
+
+@dataclass
+class NFConfiguration(Configuration):
+    """Network Function configuration."""
+
+    nf_name: str = ""
+    version: str = ""
+
+    @classmethod
+    def helptext(cls) -> "NFConfiguration":
+        """
+        Build an object where each value is helptext for that field.
+        """
+        return NFConfiguration(
+            nf_name="Name of NF definition",
+            version="Version of the NF definition in A.B.C format.",
+            **asdict(Configuration.helptext()),
+        )
+
+    def validate(self):
+        """
+        Validate the configuration.
+        """
+        super().validate()
+        if not self.nf_name:
+            raise ValidationError("nf_name must be set")
+        if not self.version:
+            raise ValidationError("version must be set")
+
+    @property
+    def nfdg_name(self) -> str:
+        """Return the NFD Group name from the NFD name."""
+        return f"{self.nf_name}-nfdg"
+
+    @property
+    def acr_manifest_names(self) -> List[str]:
+        """
+        Return the ACR manifest name from the NFD name.
+
+        This is returned in a list for consistency with the NSConfiguration, where there
+        can be multiple ACR manifests.
+        """
+        sanitized_nf_name = self.nf_name.lower().replace("_", "-")
+        return [f"{sanitized_nf_name}-acr-manifest-{self.version.replace('.', '-')}"]
+
+
+@dataclass
+class VNFConfiguration(NFConfiguration):
+    blob_artifact_store_name: str = ""
+    image_name_parameter: str = ""
+    arm_template: Union[Dict[str, str], ArtifactConfig] = ArtifactConfig()
```

Comment:
It looks like this line is not compatible with 3.11.
> Change field default mutability check, allowing only defaults which are [hashable](https://docs.python.org/3/glossary.html#term-hashable) instead of any object which is not an instance of [dict](https://docs.python.org/3/library/stdtypes.html#dict), [list](https://docs.python.org/3/library/stdtypes.html#list) or [set](https://docs.python.org/3/library/stdtypes.html#set). (Contributed by Eric V. Smith in [bpo-44674](https://bugs.python.org/issue?@action=redirect&bpo=44674).)   --[What’s New In Python 3.11 — Python 3.12.0 documentation](https://docs.python.org/3/whatsnew/3.11.html#dataclasses)

This error is raised when run `az self-test --debug`:
```
  File "/opt/az/azcliextensions/aosm/azext_aosm/_configuration.py", line 267, in <module>
    @dataclass
     ^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/dataclasses.py", line 1230, in dataclass
    return wrap(cls)
           ^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/dataclasses.py", line 1220, in wrap
    return _process_class(cls, init, repr, eq, order, unsafe_hash,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/dataclasses.py", line 958, in _process_class
    cls_fields.append(_get_field(cls, name, type, kw_only))
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/dataclasses.py", line 815, in _get_field
    raise ValueError(f'mutable default {type(f.default)} for field '
ValueError: mutable default <class 'azext_aosm._configuration.ArtifactConfig'> for field arm_template is not allowed: use default_factory
```

Reply:
Are we expected to support Python 3.11? I can't run `azdev setup` against 3.11, and the pipelines for our PR didn't test at Python 3.11, as far as I can see.

I have a potential fix, but because of the above problems, testing and ensuring there aren't other occurrences of this problem (I expect there will be), is not possible.